mesmer/core/datatree.py

```diff
@@ -0,0 +1,160 @@
+    """
+    # assert only one node in dt
+    if not len(list(dt.subtree)) == 1:
+        raise ValueError("DataTree must only contain one node.")
+    if not dt.has_data:
+        raise ValueError("DataTree must contain data.")
+
+    ds = dt.to_dataset()
+    name, *others = ds.keys()
+    if others:
+        raise ValueError("DataTree must only contain one data variable.")
+
+    da = ds.to_array().isel(variable=0).drop_vars("variable")
+    return da.rename(name)
+
+
+def collapse_datatree_into_dataset(dt: DataTree, dim: str) -> xr.Dataset:
+    """
+    Take a ``DataTree`` and collapse **all subtrees** in it into a single ``xr.Dataset`` along dim.
+    All datasets in the ``DataTree`` must have the same dimensions and each dimension must have a coordinate.
+
+    Parameters
+    ----------
+    dt : DataTree
+        The DataTree to collapse.
+    dim : str
+        The dimension to concatenate the datasets along.
+
+    Returns
+    -------
+    xr.Dataset
+        The collapsed dataset.
+
+    Raises
+    ------
+    ValueError
+        If all datasets do not have the same dimensions.
+        If any dimension does not have a coordinate.
+    """
+    # TODO: implement and test options for join and coords
+    # TODO: could potentially be replaced by DataTree.merge_child_nodes in the future?
+    datasets = [subtree.to_dataset() for subtree in dt.subtree if not subtree.is_empty]
+
+    # Check if all datasets have the same dimensions
+    first_dims = set(datasets[0].dims)
+    if not all(set(ds.dims) == first_dims for ds in datasets):
+        raise ValueError("All datasets must have the same dimensions")
+
+    # Check that all dimensions have coordinates
+    for ds in datasets:
+        for ds_dim in ds.dims:
+            if ds[ds_dim].coords == {}:
+                raise ValueError(
+                    f"Dimension '{ds_dim}' must have a coordinate/coordinates."
+                )
+
+    # Concatenate datasets along the specified dimension
+    ds = xr.concat(datasets, dim=dim)
+    ds = ds.assign_coords(
+        {dim: [subtree.name for subtree in dt.subtree if not subtree.is_empty]}
+    )
+
+    return ds
+
+
+def stack_linear_regression_datatrees(
+    predictors: DataTree,
+    target: DataTree,
+    weights: DataTree | None,
+    *,
+    stacking_dims: list[str],
+    collapse_dim: str = "scenario",
+    stacked_dim: str = "sample",
+) -> tuple[DataTree, xr.Dataset, xr.Dataset | None]:
+    """
+    prepares data for Linear Regression:
+    1. Broadcasts predictors to target
+    2. Collapses DataTrees into DataSets
+    3. Stacks the DataSets along the stacking dimension(s)
+
+    Parameters
+    ----------
+    predictors : DataTree
+        A ``DataTree`` of ``xr.Dataset`` objects used as predictors. The ``DataTree``
+        must have subtrees for each predictor each of which has to have at least one
+        leaf, holding a ``xr.Dataset`` representing a scenario. The subtrees of
```

Comment:
```suggestion
        must have subtrees for each predictor each of which has to have at least one
        non-empty leaf, representing a scenario. The subtrees of
```

maybe, not sure about this